File: slim/tools.py
"""Whitelist tools. Paths stay under data_dir. No shell.

Writes additionally go through the policy module: agent state (databases and
similar) is never writable, whoever asks, because one write_file call used to be
able to overwrite the agent's own database inside data_dir.
"""
import json
import shutil
import subprocess
import os
import urllib.error
import urllib.request

from policy import protected_path
from rag import search as rag_search

MAX_WRITE = 8 * 1024
MAX_READ = 16 * 1024
MAX_TOOL_OUTPUT = 8 * 1024
MAX_QUEUED_FILES = 16
MAX_QUEUED_BYTES = 64 * 1024


class ToolBudget(object):
    """Bounds what tool output may enter the conversation.

    Without a bound, two or three tool rounds push the prompt far past the context window
    of a small local model, and the failure looks like the model ignoring its tools.
    """

    def __init__(self, max_total=MAX_TOOL_OUTPUT):
        self.max_total = max_total
        self.used = 0
        self.elided = 0

    def add(self, text):
        """The text, trimmed to what is left, with a marker when it had to trim."""
        if self.used >= self.max_total:
            self.elided += len(text)
            return "[tool output omitted: %d-byte budget exhausted]" % self.max_total
        left = self.max_total - self.used
        if len(text) <= left:
            self.used += len(text)
            return text
        self.elided += len(text) - left
        self.used += left
        return text[:left] + "\n[truncated %d bytes: tool output budget reached]" % (len(text) - left)

    def summary(self):
        out = "tool output %d/%d bytes" % (self.used, self.max_total)
        if self.elided:
            out += ", %d bytes elided" % self.elided
        return out

    def reset(self):
        self.used = 0
        self.elided = 0


class ChangeQueue(object):
    """A turn's planned file changes.

    Staging instead of writing gives the operator one manifest for the whole turn, and
    keeps a model that changes its mind from leaving half a plan on disk. Last write to
    the same path wins, and that is reported.
    """

    def __init__(self, max_files=MAX_QUEUED_FILES, max_bytes=MAX_QUEUED_BYTES):
        self.max_files = max_files
        self.max_bytes = max_bytes
        self.entries = []
        self.replaced = 0

    def stage(self, rel, full, content):
        if len(content) > MAX_WRITE:
            raise ValueError("refusing to stage %s: %d bytes is larger than the %d-byte limit"
                             % (rel, len(content), MAX_WRITE))
        for entry in self.entries:
            if entry["full"] == full:
                entry["content"] = content
                self.replaced += 1
                return
        total = sum(len(e["content"]) for e in self.entries)
        if len(self.entries) >= self.max_files:
            raise ValueError("refusing to stage %s: the change queue already holds %d files"
                             % (rel, self.max_files))
        if total + len(content) > self.max_bytes:
            raise ValueError("refusing to stage %s: the change queue would exceed %d bytes"
                             % (rel, self.max_bytes))
        self.entries.append({"rel": rel, "full": full, "content": content})

    def apply(self):
        """Write every staged file. Returns (wrote, failures); one failure never discards
        the rest."""
        wrote, failures = [], []
        for entry in self.entries:
            parent = os.path.dirname(entry["full"])
            try:
                if parent and not os.path.isdir(parent):
                    os.makedirs(parent)
                with open(entry["full"], "w", encoding="utf-8") as fh:
                    fh.write(entry["content"])
            except OSError as exc:
                failures.append("%s: %s" % (entry["rel"], exc))
                continue
            wrote.append("wrote %s (%d bytes)" % (entry["rel"], len(entry["content"])))
        return wrote, failures

    def summary(self):
        total = sum(len(e["content"]) for e in self.entries)
        out = "%d file(s), %d bytes" % (len(self.entries), total)
        if self.replaced:
            out += ", %d overwrite(s) of an earlier staged change" % self.replaced
        return out

    def clear(self):
        self.entries = []
        self.replaced = 0

TOOLS = (
    "rag_search: {\"tool\":\"rag_search\",\"query\":\"keywords\"}",
    "read_file: {\"tool\":\"read_file\",\"path\":\"relative.md\"}",
    "write_file: {\"tool\":\"write_file\",\"path\":\"note.txt\",\"content\":\"...\"}",
    "mqtt_publish: {\"tool\":\"mqtt_publish\",\"topic\":\"a/b\",\"payload\":\"hi\"}  (needs SLIM_MQTT_HTTP)",
    "run_command: {\"tool\":\"run_command\",\"command\":\"uptime\",\"args\":\"-s\"}  (needs --allow-exec and a name in commands.allow)",
)


def _jail(data_dir, rel):
    if not rel or os.path.isabs(rel) or ".." in rel.replace("\\", "/").split("/"):
        raise ValueError("path must be relative and stay in data dir")
    full = os.path.normpath(os.path.join(data_dir, rel))
    root = os.path.normpath(data_dir)
    if full != root and not full.startswith(root + os.sep):
        raise ValueError("path escapes data dir")
    return full


EXEC_TIMEOUT = 10          # seconds; a runaway command must not hold the turn open
MAX_EXEC_OUTPUT = 16 * 1024
DEFAULT_EXEC_PATH = "/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin"


def _bare_command_name(name):
    """No path, no `.`/`..`: a model must not point the runner at something it wrote into
    the data directory."""
    return bool(name) and name not in (".", "..") and "/" not in name


def _bounded(budget, text):
    """Every successful tool result passes the same budget so no tool can flood the
    conversation. Errors are returned unbudgeted on purpose: they are short, and they are
    exactly what the model needs to see."""
    return budget.add(text) if budget else text


def run_tool(name, args, conn, data_dir, mqtt_http="", queue=None, budget=None,
             exec_allow=None, exec_path=None):
    args = args or {}
    if name == "rag_search":
        hits = rag_search(conn, args.get("query") or "", limit=int(args.get("limit") or 3))
        return _bounded(budget, json.dumps(hits, ensure_ascii=False))
    if name == "read_file":
        full = _jail(data_dir, args.get("path") or "")
        with open(full, "r", encoding="utf-8", errors="replace") as fh:
            return _bounded(budget, fh.read(MAX_READ))
    if name == "write_file":
        rel = args.get("path") or ""
        full = _jail(data_dir, rel)
        reason = protected_path(rel)
        if reason:
            raise ValueError(reason)
        content = args.get("content") or ""
        if len(content) > MAX_WRITE:
            raise ValueError("write over %s bytes" % MAX_WRITE)
        if queue is not None:
            # Staged, not written: the queue is applied once when the turn ends, and the
            # manifest is what the operator audits.
            queue.stage(rel, full, content)
            return _bounded(budget, "staged %s (%d bytes); it is written when this turn ends"
                            % (rel, len(content)))
        parent = os.path.dirname(full)
        if parent and not os.path.isdir(parent):
            os.makedirs(parent)
        with open(full, "w", encoding="utf-8") as fh:
            fh.write(content)
        return "wrote %s (%s bytes)" % (args.get("path"), len(content))
    if name == "mqtt_publish":
        if not mqtt_http:
            raise ValueError("mqtt disabled; set SLIM_MQTT_HTTP to an HTTP ingest URL")
        payload = json.dumps(
            {"topic": args.get("topic") or "", "payload": args.get("payload") or ""},
            ensure_ascii=False,
        ).encode("utf-8")
        req = urllib.request.Request(
            mqtt_http,
            data=payload,
            headers={"Content-Type": "application/json", "Content-Length": str(len(payload))},
            method="POST",
        )
        try:
            resp = urllib.request.urlopen(req, timeout=10)
            try:
                body = resp.read(2048)
            finally:
                resp.close()
        except urllib.error.URLError as exc:
            raise ValueError("mqtt http failed: %s" % exc.reason)
        return _bounded(budget, "mqtt http %s" % body[:200].decode("utf-8", "replace"))
    if name == "run_command":
        # Three gates, same order as the C++ side: a bare name, the allowlist (a missing
        # file permits nothing), and the model opt-in, which approve_tool() already applied
        # before this call. shell is never involved - argv reaches the program as-is.
        cmd = (args.get("command") or "").strip()
        if not _bare_command_name(cmd):
            raise ValueError("refusing to run '%s': a bare command name is required "
                             "(no path, no directory part)" % cmd)
        if cmd not in list(exec_allow or []):
            raise ValueError("refusing to run %s: it is not in the command allowlist "
                             "(<data-dir>/commands.allow)" % cmd)
        path = shutil.which(cmd, path=exec_path or DEFAULT_EXEC_PATH)
        if not path:
            raise ValueError("command not found: %s" % cmd)
        argv = [path] + str(args.get("args") or "").split()
        try:
            proc = subprocess.run(argv, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE,
                                  stderr=subprocess.STDOUT, timeout=EXEC_TIMEOUT)
        except subprocess.TimeoutExpired:
            raise ValueError("command timed out after %ds" % EXEC_TIMEOUT)
        raw = proc.stdout or b""
        out = raw[:MAX_EXEC_OUTPUT].decode("utf-8", "replace").rstrip("\n")
        if len(raw) > MAX_EXEC_OUTPUT:
            out += "\n[output truncated at %d bytes]" % MAX_EXEC_OUTPUT
        return _bounded(budget, "exit %d\n%s" % (proc.returncode, out))
    raise ValueError("unknown tool: %s" % name)


def parse_tool_call(text):
    """Return (name, args) if the model emitted a JSON tool object, else None."""
    if not text:
        return None
    start = text.find("{")
    end = text.rfind("}")
    if start < 0 or end <= start:
        return None
    try:
        obj = json.loads(text[start : end + 1])
    except ValueError:
        return None
    if not isinstance(obj, dict):
        return None
    name = obj.get("tool")
    if not name:
        return None
    return name, obj

File: slim/cpp/tools.cpp
#include "tools.hpp"

#include "http.hpp"
#include "util.hpp"

#include <cstdio>
#include <iostream>
#include <sstream>

namespace {

void audit_tool(const std::string& tool, const std::string& detail, bool allowed,
                const std::string& why) {
  std::cerr << "[slim] tool " << tool << " (" << detail << ") -> " << (allowed ? "allow" : "deny")
            << ": " << why << "\n";
}

std::string int_str(long n) {
  std::ostringstream o;
  o << n;
  return o.str();
}

std::string number(size_t n) {
  std::ostringstream o;
  o << n;
  return o.str();
}

}  // namespace

std::string ToolBudget::add(const std::string& text) {
  if (used >= max_total_bytes) {
    elided += text.size();
    return "[tool output omitted: " + number(max_total_bytes) + "-byte budget exhausted]";
  }
  size_t left = max_total_bytes - used;
  if (text.size() <= left) {
    used += text.size();
    return text;
  }
  elided += text.size() - left;
  used += left;
  return text.substr(0, left) + "\n[truncated " + number(text.size() - left) +
         " bytes: tool output budget reached]";
}

std::string ToolBudget::summary() const {
  std::string s = "tool output " + number(used) + "/" + number(max_total_bytes) + " bytes";
  if (elided)
    s += ", " + number(elided) + " bytes elided";
  return s;
}

void ToolBudget::reset() {
  used = 0;
  elided = 0;
}

bool ChangeQueue::stage(const std::string& rel, const std::string& full, const std::string& content,
                        std::string& why) {
  if (content.size() > kMaxWrite) {
    why = "refusing to stage " + rel + ": " + number(content.size()) + " bytes is larger than the " +
          number(kMaxWrite) + "-byte limit for one file";
    return false;
  }
  size_t total = 0;
  for (size_t i = 0; i < entries.size(); ++i) {
    if (entries[i].full == full) {
      // Last write wins, but say so: a model that changes its mind about a file is worth
      // seeing in the manifest.
      entries[i].content = content;
      ++replaced;
      return true;
    }
    total += entries[i].content.size();
  }
  if (entries.size() >= kMaxQueuedFiles) {
    why = "refusing to stage " + rel + ": the change queue already holds " +
          number(kMaxQueuedFiles) + " files";
    return false;
  }
  if (total + content.size() > kMaxQueuedBytes) {
    why = "refusing to stage " + rel + ": the change queue would exceed " +
          number(kMaxQueuedBytes) + " bytes";
    return false;
  }
  Entry e;
  e.rel = rel;
  e.full = full;
  e.content = content;
  entries.push_back(e);
  return true;
}

std::vector<std::string> ChangeQueue::apply(std::vector<std::string>& failures) {
  std::vector<std::string> wrote;
  for (size_t i = 0; i < entries.size(); ++i) {
    std::string err;
    if (!write_file_limited(entries[i].full, entries[i].content, kMaxWrite, err)) {
      failures.push_back(entries[i].rel + ": " + err);
      continue;
    }
    wrote.push_back("wrote " + entries[i].rel + " (" + number(entries[i].content.size()) + " bytes)");
  }
  return wrote;
}

std::string ChangeQueue::summary() const {
  size_t total = 0;
  for (size_t i = 0; i < entries.size(); ++i)
    total += entries[i].content.size();
  std::string s = number(entries.size()) + " file(s), " + number(total) + " bytes";
  if (replaced)
    s += ", " + number(replaced) + " overwrite(s) of an earlier staged change";
  return s;
}

void ChangeQueue::clear() {
  entries.clear();
  replaced = 0;
}

bool known_tool(const std::string& name) {
  return name == "rag_search" || name == "read_file" || name == "write_file" ||
         name == "mqtt_publish" || name == "run_command";
}

bool parse_tool_call(const std::string& text, std::string& name, std::string& json_obj) {
  size_t start = text.find('{');
  if (start == std::string::npos)
    return false;
  size_t end = text.rfind('}');
  if (end <= start)
    return false;
  json_obj = text.substr(start, end - start + 1);
  if (!json_extract_string(json_obj, "tool", name))
    return false;
  return known_tool(name);
}

std::string run_tool(const ToolEnv& env, const std::string& name, const std::string& obj,
                     bool human_initiated, ToolBudget* budget, ChangeQueue* queue) {
  std::string result;
  if (name == "rag_search") {
    if (!env.rag_fn)
      return "tool error: rag search is not available in this build";
    std::string q;
    json_get_string_field(obj, "query", q);
    result = env.rag_fn(env.rag_ctx, q);
  } else if (name == "read_file") {
    std::string rel, full, err;
    json_get_string_field(obj, "path", rel);
    if (!jail_path(env.data_dir, rel, full, err))
      return "tool error: " + err;
    result = read_file_limited(full, 16 * 1024);
    if (result.empty())
      return "tool error: empty or missing";
  } else if (name == "write_file") {
    std::string rel, content, full, err, why;
    json_get_string_field(obj, "path", rel);
    json_get_string_field(obj, "content", content);
    if (!jail_path(env.data_dir, rel, full, err))
      return "tool error: " + err;
    if (protected_path(rel, why)) {
      audit_tool(name, rel, false, why);
      return "tool error: " + why;
    }
    if (!approve_mutation(env.policy, name, rel, human_initiated, why)) {
      audit_tool(name, rel, false, why);
      return "tool error: " + why;
    }
    if (queue) {
      // Staged, not written: the queue is applied once when the turn ends, and the
      // manifest is what the operator audits.
      if (!queue->stage(rel, full, content, why)) {
        audit_tool(name, rel, false, why);
        return "tool error: " + why;
      }
      audit_tool(name, rel, true, why + " (staged until the end of the turn)");
      return "staged " + rel + " (" + number(content.size()) +
             " bytes); it is written when this turn ends";
    }
    audit_tool(name, rel, true, why);
    if (!write_file_limited(full, content, kMaxWrite, err))
      return "tool error: " + err;
    result = "wrote " + rel;
  } else if (name == "mqtt_publish") {
    if (env.mqtt_http.empty())
      return "tool error: mqtt disabled; set SLIM_MQTT_HTTP";
    std::string topic, payload, why;
    json_get_string_field(obj, "topic", topic);
    json_get_string_field(obj, "payload", payload);
    if (!approve_mutation(env.policy, name, topic, human_initiated, why)) {
      audit_tool(name, topic, false, why);
      return "tool error: " + why;
    }
    audit_tool(name, topic, true, why);
    std::ostringstream js;
    js << "{\"topic\":\"" << json_escape(topic) << "\",\"payload\":\"" << json_escape(payload)
       << "\"}";
    // Deliberately single-shot: a publish is a mutation, and a retry after an ambiguous
    // failure can duplicate the effect. The C++ test counts requests on a local server.
    HttpResult r = http_post_json(env.mqtt_http, js.str(), "", 10);
    if (!r.error.empty() && r.status == 0)
      return "tool error: " + r.error;
    result = "mqtt http " + r.body.substr(0, 200);
  } else if (name == "run_command") {
    // Execution is the one tool that changes the machine, so it has three gates, in order:
    // a bare command name (no paths - a model must not point this at a binary it wrote),
    // membership in the allowlist (a missing or empty file permits nothing), and the usual
    // model-initiated approval. No shell is involved at any point in this path.
    std::string cmd, argstr, why;
    json_get_string_field(obj, "command", cmd);
    json_get_string_field(obj, "args", argstr);
    if (!bare_command_name(cmd)) {
      audit_tool(name, cmd, false, "refusing a command that is not a bare name");
      return "tool error: refusing to run '" + cmd +
             "': a bare command name is required (no path, no directory part)";
    }
    if (!string_in_list(env.policy.exec_allow, cmd)) {
      audit_tool(name, cmd, false, "not in the command allowlist");
      return "tool error: refusing to run " + cmd +
             ": it is not in the command allowlist (<data-dir>/commands.allow)";
    }
    if (!approve_mutation(env.policy, name, cmd, human_initiated, why)) {
      audit_tool(name, cmd, false, why);
      return "tool error: " + why;
    }
    // Whitespace-split arguments: no quoting, no globbing, no shell. Each token is one
    // argv entry, so `;`, `|` or `$(...)` inside a token is just a character.
    std::vector<std::string> argv;
    argv.push_back(cmd);
    {
      std::string cur;
      for (size_t i = 0; i <= argstr.size(); ++i) {
        if (i == argstr.size() || argstr[i] == ' ' || argstr[i] == '\t') {
          if (!cur.empty()) {
            argv.push_back(cur);
            cur.clear();
          }
        } else {
          cur.push_back(argstr[i]);
        }
      }
    }
    audit_tool(name, cmd + (argstr.empty() ? std::string() : " " + argstr), true, why);
    std::string out, err;
    int status = 0;
    if (!run_argv_capture(argv, env.exec_path, kExecTimeoutSec, kMaxExecOutput, out, status, err)) {
      audit_tool(name, cmd, false, err);
      return "tool error: " + err;
    }
    result = "exit " + int_str(status) + "\n" + out;
  } else {
    // Unknown tool: refused rather than ignored, so a typo or a new tool name cannot
    // silently pass through.
    audit_tool(name, "", false, "not a classified tool (fail-closed)");
    return "tool error: unknown tool";
  }
  if (budget)
    return budget->add(result);
  return result;
}
